File: data_structures/5_kotlin/l_hashMaps.kt
data class KeyValue<T : Any, U>(val key: T, var value: U)

class HashMap<T : Any, U>(capacity: Int) {
    private val buckets: Array<MutableList<KeyValue<T, U>>>
    private val capacity: Int

    init {
        this.capacity = maxOf(1, capacity)
        buckets = Array(this.capacity) { mutableListOf() }
    }

    // Insertion
    fun setValue(value: U, key: T) {
        val index = bucketIndex(key)
        for ((i, kv) in buckets[index].withIndex()) {
            if (kv.key == key) {
                buckets[index][i].value = value
                return
            }
        }
        buckets[index].add(KeyValue(key, value))
    }

    // Retrieval
    fun getValue(key: T): U? {
        val index = bucketIndex(key)
        for (kv in buckets[index]) {
            if (kv.key == key) {
                return kv.value
            }
        }
        return null
    }

    // Removal
    fun removeValue(key: T) {
        val index = bucketIndex(key)
        buckets[index].removeIf { it.key == key }
    }

    // Count
    fun count(): Int {
        return buckets.sumBy { it.size }
    }

    // Keys and Values
    fun allKeys(): List<T> {
        return buckets.flatMap { it.map { kv -> kv.key } }
    }

    fun allValues(): List<U> {
        return buckets.flatMap { it.map { kv -> kv.value } }
    }

    // Helper Functions
    fun isEmpty(): Boolean {
        return buckets.all { it.isEmpty() }
    }

    fun removeAll() {
        buckets.forEach { it.clear() }
    }

    fun contains(key: T): Boolean {
        val index = bucketIndex(key)
        return buckets[index].any { it.key == key }
    }

    fun loadFactor(): Double {
        return count().toDouble() / capacity.toDouble()
    }

    fun keyValuePairs(): Set<KeyValue<T, U>> {
        return buckets.flatMapTo(mutableSetOf()) { it }
    }

    // Update Values
    fun updateValue(value: U, key: T) {
        val index = bucketIndex(key)
        for ((i, kv) in buckets[index].withIndex()) {
            if (kv.key == key) {
                buckets[index][i].value = value
                return
            }
        }
        buckets[index].add(KeyValue(key, value))
    }

    fun updateValues(dictionary: Map<T, U>) {
        for ((key, value) in dictionary) {
            updateValue(value, key)
        }
    }

    // ... Add more hashmap operations as needed ...

    private fun bucketIndex(key: T): Int {
        // Simplified hash function for demonstration purposes
        return key.hashCode() % capacity
    }
}

fun main() {
    val hashmap = HashMap<String, Int>(10)
    hashmap.setValue(5, "five")
    hashmap.setValue(10, "ten")

    println("Value for key 'five': ${hashmap.getValue("five")}")
    println("Value for key 'ten': ${hashmap.getValue("ten")}")

    hashmap.removeValue("five")
    println("Contains key 'five': ${hashmap.contains("five")}")
    println("Total count: ${hashmap.count()}")

    val keys = hashmap.allKeys()
    println("All keys: $keys")

    val values = hashmap.allValues()
    println("All values: $values")

    println("Is empty: ${hashmap.isEmpty()}")
    println("Load factor: ${hashmap.loadFactor()}")
}

File: data_structures/3_python/l_hashMaps.py
class KeyValue:
    def __init__(self, key, value):
        self.key = key
        self.value = value


class HashMap:
    def __init__(self, capacity):
        self.capacity = max(1, capacity)
        self.buckets = [[] for _ in range(self.capacity)]

    # Insertion
    def set_value(self, value, key):
        index = self.bucket_index(key)
        for i, kv in enumerate(self.buckets[index]):
            if kv.key == key:
                self.buckets[index][i].value = value
                return
        self.buckets[index].append(KeyValue(key, value))

    # Retrieval
    def get_value(self, key):
        index = self.bucket_index(key)
        for kv in self.buckets[index]:
            if kv.key == key:
                return kv.value
        return None

    # Removal
    def remove_value(self, key):
        index = self.bucket_index(key)
        self.buckets[index] = [kv for kv in self.buckets[index] if kv.key != key]

    # Count
    def count(self):
        return sum(len(bucket) for bucket in self.buckets)

    # Keys and Values
    def all_keys(self):
        return [kv.key for bucket in self.buckets for kv in bucket]

    def all_values(self):
        return [kv.value for bucket in self.buckets for kv in bucket]

    # Helper Functions
    def is_empty(self):
        return all(not bucket for bucket in self.buckets)

    def remove_all(self):
        self.buckets = [[] for _ in range(self.capacity)]

    def contains(self, key):
        index = self.bucket_index(key)
        return any(kv.key == key for kv in self.buckets[index])

    def load_factor(self):
        return self.count() / self.capacity

    def key_value_pairs(self):
        return {kv.key: kv.value for bucket in self.buckets for kv in bucket}

    # Update Values
    def update_value(self, value, key):
        index = self.bucket_index(key)
        for i, kv in enumerate(self.buckets[index]):
            if kv.key == key:
                self.buckets[index][i].value = value
                return
        self.buckets[index].append(KeyValue(key, value))

    def update_values(self, dictionary):
        for key, value in dictionary.items():
            self.update_value(value, key)

    # ... Add more hashmap operations as needed ...

    def bucket_index(self, key):
        # Simplified hash function for demonstration purposes
        return hash(key) % self.capacity


if __name__ == "__main__":
    hashmap = HashMap(10)
    hashmap.set_value(5, "five")
    hashmap.set_value(10, "ten")

    print("Value for key 'five':", hashmap.get_value("five"))
    print("Value for key 'ten':", hashmap.get_value("ten"))

    hashmap.remove_value("five")
    print("Contains key 'five':", hashmap.contains("five"))
    print("Total count:", hashmap.count())

    keys = hashmap.all_keys()
    print("All keys:", keys)

    values = hashmap.all_values()
    print("All values:", values)

    print("Is empty:", hashmap.is_empty())
    print("Load factor:", hashmap.load_factor())
